Refactored RqForm into 1 interface and 3 smaller classes.

rui-castro opened the pull request:
Removed checkstyle suppression ClassDataAbstractionCouplingCheck from RqForm, but is still needed in RqFormBase.

davvd commented:
@rui-castro Many thanks, I will find someone to review it before we merge

davvd commented:
@mkordas this pull request is for you, please review

mkordas commented:
@rui-castro thanks for PR, I'll do a review tomorrow

mkordas commented:
@rui-castro can you remove word `closes` from PR description? Only author of issue can close it, we cannot let GitHub do it.

mkordas commented on src/main/java/org/takes/rq/form/package-info.java:
```diff
@@ -0,0 +1,32 @@
+
+/**
+ * Implementations of the interface {@link org.takes.rq.RqForm}.
+ *
```
@rui-castro see Yegor's comment in https://github.com/teamed/qulice/issues/708#issuecomment-184518026

rui-castro commented:
@mkordas Removed closed from PR description and fixed comment formatting.

mkordas commented:
@rui-castro thanks, I will continue my review soon

mkordas commented on src/main/java/org/takes/rq/form/RqFormBase.java:
```diff
@@ -0,0 +1,177 @@
+ * @author Aleksey Popov (alopen@yandex.ru)
+ * @version $Id$
+ * @since 0.9
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
```
@rui-castro looks like issue is not fully fixed, we need to create additional puzzle for it

rui-castro replied:
@mkordas yes, I can add a todo but I don't know how to create a puzzle.

mkordas replied:
@rui-castro just add `@todo` as in other places :)

mkordas commented on src/main/java/org/takes/rq/form/RqFormBase.java:
```diff
@@ -0,0 +1,177 @@
+ * Base implementation of @link RqForm.
+ * @author Aleksey Popov (alopen@yandex.ru)
+ * @version $Id$
+ * @since 0.9
```
@rui-castro this class is new, should be `0.33`

rui-castro replied:
@mkordas The @author and the @since tags were copied from the original inner classes. Should I update them? If yes, should I update both or just the @since tag?

mkordas replied:
@rui-castro at least `@since` but if you want you can also add you as an author

mkordas commented on src/main/java/org/takes/rq/form/RqFormFake.java:
```diff
@@ -0,0 +1,125 @@
+ * RqFormFake accepts parameters in the constructor.
+ * @author Erim Erturk (erimerturk@gmail.com)
+ * @version $Id$
+ * @since 0.22
```
@rui-castro same here, we need to update since

mkordas commented on src/main/java/org/takes/rq/form/RqFormSmart.java:
```diff
@@ -0,0 +1,117 @@
+ *
+ * @author Yegor Bugayenko (yegor@teamed.io)
+ * @version $Id$
+ * @since 0.14
```
@rui-castro and here as well

mkordas commented on src/test/java/org/takes/rq/form/RqFormBaseTest.java:
```diff
@@ -0,0 +1,102 @@
+
+/**
+ * Test case for {@link RqFormBase}.
+ * @author Yegor Bugayenko (yegor@teamed.io)
```
@rui-castro it cannot be Yegor :)

rui-castro replied:
@mkordas Yegor is the author of the original class https://github.com/yegor256/takes/blob/master/src/test/java/org/takes/rq/RqFormTest.java#L36.
I simply splitted the class in 3 new classes. Should I change the @author to myself?

mkordas replied:
@rui-castro we can have two authors

mkordas replied:
@rui-castro it just looks strange as Git shows it as brand new file, so it cannot have old autor as the only author

mkordas commented on src/test/java/org/takes/rq/form/RqFormBaseTest.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    /**
+     * Returns always same instances (Cache).
```
@rui-castro `RqFormBase can` is missing here

mkordas commented on src/test/java/org/takes/rq/form/RqFormBaseTest.java:
```diff
@@ -0,0 +1,102 @@
+    private static final String HEADER = "Content-Length: %d";
+
+    /**
+     * RqForm can parse body.
```
@rui-castro `RqForm` -> `RqFormBase`

mkordas commented on src/main/java/org/takes/rq/form/package-info.java:
```diff
@@ -0,0 +1,31 @@
+ */
+
+/**
+ * Implementations of the interface {@link org.takes.rq.RqForm}.
```
@rui-castro I think that's the place where we should put more info what `form` is

rui-castro replied:
@mkordas Honestly, I have no idea about what relevant info I could add here. Besides that, writing documentation was not the purpose of this issue, right? :)

mkordas replied:
@rui-castro but we created a new package that should be nicely described. We can e.g. create 15 min puzzle for it.

rui-castro commented:
@mkordas I think I addressed all your comments with my last commit.

mkordas commented:
@rui-castro I'll check it soon

mkordas commented on src/test/java/org/takes/rq/form/RqFormBaseTest.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    /**
+     * RqFormBase returns always the same instances (Cache).
```
@rui-castro second word should be `can`

mkordas commented on src/main/java/org/takes/rq/form/RqFormBase.java:
```diff
@@ -0,0 +1,180 @@
+ * @version $Id$
+ * @since 0.33
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ * @todo Refactor this class to reduce the data abstraction
```
@rui-castro you lack issue number and time needed

mkordas commented:
@rui-castro Travis build has failed

mkordas commented:
@rui-castro see my review above

mkordas commented:
@rui-castro can you please solve conflicts here?

rui-castro commented:
@mkordas Merge conflicts solved and comments addressed. 
The travis build failed because package org.takes.rq.form has a short description, but I already created the puzzle to improve it.

mkordas commented:
@rui-castro I'll check it soon

mkordas commented:
@rui-castro we cannot have failed build. We need to fix the puzzle:
```
ERROR: "puzzle src/main/java/org/takes/rq/form/package-info.java:30-30 has a very short description of just 5 words while a minimum of 20 is required"
/home/travis/.rvm/gems/ruby-1.9.3-p551/gems/pdd-0.15.4/lib/pdd.rb:150:in `rules': 1 errors, see log above (PDD::Error)
	from /home/travis/.rvm/gems/ruby-1.9.3-p551/gems/pdd-0.15.4/lib/pdd.rb:105:in `xml'
	from /home/travis/.rvm/gems/ruby-1.9.3-p551/gems/pdd-0.15.4/bin/pdd:90:in `<top (required)>'
	from /home/travis/.rvm/gems/ruby-1.9.3-p551/bin/pdd:23:in `load'
	from /home/travis/.rvm/gems/ruby-1.9.3-p551/bin/pdd:23:in `<main>'
	from /home/travis/.rvm/gems/ruby-1.9.3-p551/bin/ruby_executable_hooks:15:in `eval'
	from /home/travis/.rvm/gems/ruby-1.9.3-p551/bin/ruby_executable_hooks:15:in `<main>'
```

mkordas commented on src/main/java/org/takes/rq/form/RqFormBase.java:
```diff
@@ -0,0 +1,180 @@
+ * @version $Id$
+ * @since 0.33
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ * @todo #687:15min Refactor this class to reduce the data abstraction
```
@rui-castro I think we should make it 30 min, it's not that easy

mkordas commented on src/main/java/org/takes/rq/form/package-info.java:
```diff
@@ -0,0 +1,32 @@
+ * @author Rui Castro (rui.castro@gmail.com)
+ * @version $Id$
+ * @since 0.33
+ * @todo #687:15min Improve documentation for package org.takes.rq.form.
```
@rui-castro we need more information for task performer here, some context etc.

mkordas commented:
@rui-castro see my comments above

rui-castro commented:
@mkordas I've addressed your comments with the last commit.
I've improved the description for puzzle to improve the package documentation, but I'm not entirely sure if that's what you want. If you think it's still not enough, and since you were the one who proposed the improvement of documentation, please advise on what I should write in there.

mkordas commented:
@rui-castro I'll check it soon

mkordas commented:
@rui-castro Travis build is failed, we need to fix it

mkordas commented:
@rui-castro I think that description of puzzle is still too short:
```
ERROR: "puzzle src/main/java/org/takes/rq/form/package-info.java:36-38 has a very short description of just 19 words while a minimum of 20 is required"

mkordas commented:
@rui-castro see my comment above

rui-castro commented:
@mkordas the build is fixed.

mkordas commented:
@rui-castro I'll check it in a sec :)

mkordas commented:
@rui-castro sorry for a delay, on it now

mkordas commented:
@rui-castro looks good!

mkordas commented:
@rultor merge

rultor commented:
> @rultor merge

@mkordas Thanks for your request. @yegor256 Please confirm this.

rui-castro commented:
@mkordas thanks 😉

mkordas commented:
@yegor256 we need your review here

yegor256 commented:
@rultor try to merge

rultor commented:
> @rultor try to merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/8925-229218409)

rultor commented:
> @rultor try to merge

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/8925-229218409) (took me 37min)

davvd commented:
@ypshenychka please, review this issue for QA compliance, as per [par.24](http://at.teamed.io/policy.html#24)

davvd commented:
@rultor please deploy

rultor commented:
> @rultor please deploy

@davvd OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/8925-229631265)

rultor commented:
> @rultor please deploy

@davvd Done! FYI, the full log is [here](http://www.rultor.com/t/8925-229631265) (took me 33min)

ypshenychka commented:
@rui-castro 
According to our [QA Rules](http://at.teamed.io/qa.html):
>Pull request description explains the solution proposed and contains a link to the original ticket it is related to.

Please provide a link to original ticket in PR's description.

rui-castro commented:
@ypshenychka done

ypshenychka commented:
@rui-castro Thank you.

ypshenychka commented:
@davvd Quality is acceptable.

davvd commented:
> @davvd Quality is acceptable.

@ypshenychka thank you, let's try to make it "good" next time

davvd commented:
@mkordas 10 mins was added to the account of @ypshenychka (for QA review), in transaction 91761879. Thank you, I have added **52 mins** to your account in payment/transaction "`577647ae80c2b22187000434`", time consumed: 216 hours and 8 mins. you're getting extra minutes here (c=37). +52 added to your rating, at the moment it is: [+6236](http://www.netbout.com/b/36080?open=rating)
